src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -178,8 +262,362 @@ class IonCursorBinary implements IonCursor {
+     * FIXED_SIZE_CONFIGURATIONS(logBase2(size)). Only supports sizes less than or equal to
+     * STANDARD_BUFFER_CONFIGURATION.getInitialBufferSize().
+     */
+    private static final IonBufferConfiguration[] FIXED_SIZE_CONFIGURATIONS;
+
+    static {
+        int maxBufferSizeExponent = logBase2(STANDARD_BUFFER_CONFIGURATION.getInitialBufferSize());
+        FIXED_SIZE_CONFIGURATIONS = new IonBufferConfiguration[maxBufferSizeExponent + 1];
+        for (int i = 0; i <= maxBufferSizeExponent; i++) {
+            // Create a buffer configuration for buffers of size 2^i. The minimum size is 8: the smallest power of two
+            // larger than the minimum buffer size allowed.
+            int size = Math.max(8, (int) Math.pow(2, i));
+            FIXED_SIZE_CONFIGURATIONS[i] = IonBufferConfiguration.Builder.from(STANDARD_BUFFER_CONFIGURATION)
+                .withInitialBufferSize(size)
+                .withMaximumBufferSize(size)
+                .build();
+        }
+    }
+
+    /**
+     * Validates the given configuration.
+     * @param configuration the configuration to validate.
+     * @return the validated configuration.
+     */
+    private static IonBufferConfiguration validate(IonBufferConfiguration configuration) {
+        if (configuration.getInitialBufferSize() < 1) {
+            throw new IllegalArgumentException("Initial buffer size must be at least 1.");
+        }
+        if (configuration.getMaximumBufferSize() < configuration.getInitialBufferSize()) {
+            throw new IllegalArgumentException("Maximum buffer size cannot be less than the initial buffer size.");
+        }
+        return configuration;
+    }
+
+    /**
+     * Constructs a refillable cursor from the given input stream.
+     * @param configuration the configuration to use.
+     * @param alreadyRead the byte array containing the bytes already read (often the IVM).
+     * @param alreadyReadOff the offset into `alreadyRead` at which the first byte that was already read exists.
+     * @param alreadyReadLen the number of bytes already read from `alreadyRead`.
+     */
+    IonCursorBinary(
+        IonBufferConfiguration configuration,
+        InputStream inputStream,
+        byte[] alreadyRead,
+        int alreadyReadOff,
+        int alreadyReadLen
+    ) {
+        this.dataHandler = (configuration == null) ? null : configuration.getDataHandler();
+        if (configuration == null) {
+            if (inputStream instanceof ByteArrayInputStream) {
+                // ByteArrayInputStreams are fixed-size streams. Clamp the reader's internal buffer size at the size of
+                // the stream to avoid wastefully allocating extra space that will never be needed. It is still
+                // preferable for the user to manually specify the buffer size if it's less than the default, as doing
+                // so allows this branch to be skipped.
+                int fixedBufferSize;
```

Comment:
Maybe? Would it just be clearer to say you can construct an `IonCursorBinary` from a config _or_ a BAIS, with different constructors or what have you.

Reply:
I'm not sure I understand the "config or BAIS" part of your comment, as the config and stream are both used; we only set a different config if one is not provided. As for having a different constructor for ByteArrayInputStream, that is possible, but I don't believe it actually simplifies the code, as this branch contains the only non-duplicate logic between the two. And, since the higher-level abstractions do not special-case ByteArrayInputStream, we'd still need the `instanceof` check to happen somewhere. If you meant something else, let me know.

What I can do is factor out the handling into ByteArrayInputStream into a separate method to hopefully aid readability.

Reply:
> What I can do is factor out the handling into ByteArrayInputStream into a separate method to hopefully aid readability.

This is done in https://github.com/amazon-ion/ion-java/pull/538/files